core/src/mast/serialization/info.rs

```diff
@@ -0,0 +1,456 @@
+
+                Self::Block { len }
+            }
+            Join(join_node) => Self::Join {
+                left_child_id: join_node.first().0,
+                right_child_id: join_node.second().0,
+            },
+            Split(split_node) => Self::Split {
+                if_branch_id: split_node.on_true().0,
+                else_branch_id: split_node.on_false().0,
+            },
+            Loop(loop_node) => Self::Loop {
+                body_id: loop_node.body().0,
+            },
+            Call(call_node) => {
+                if call_node.is_syscall() {
+                    Self::SysCall {
+                        callee_id: call_node.callee().0,
+                    }
+                } else {
+                    Self::Call {
+                        callee_id: call_node.callee().0,
+                    }
+                }
+            }
+            Dyn => Self::Dyn,
+            External(_) => Self::External,
+        }
+    }
+}
+
+impl Serializable for MastNodeType {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        let serialized_bytes = {
+            let mut serialized_bytes = self.inline_data_to_bytes();
+
+            // Tag is always placed in the first four bytes
+            let discriminant = self.discriminant();
+            assert!(discriminant <= 0b1111);
+            serialized_bytes[0] |= discriminant << 4;
+
+            serialized_bytes
+        };
+
+        serialized_bytes.write_into(target)
+    }
+}
+
+/// Serialization helpers
+impl MastNodeType {
+    fn discriminant(&self) -> u8 {
+        // SAFETY: This is safe because we have given this enum a primitive representation with
+        // #[repr(u8)], with the first field of the underlying union-of-structs the discriminant.
+        //
+        // See the section on "accessing the numeric value of the discriminant"
+        // here: https://doc.rust-lang.org/std/mem/fn.discriminant.html
+        unsafe { *<*const _>::from(self).cast::<u8>() }
+    }
+
+    fn inline_data_to_bytes(&self) -> [u8; 8] {
+        match self {
+            MastNodeType::Join {
+                left_child_id: left,
+                right_child_id: right,
+            } => Self::encode_join_or_split(*left, *right),
+            MastNodeType::Split {
+                if_branch_id: if_branch,
+                else_branch_id: else_branch,
+            } => Self::encode_join_or_split(*if_branch, *else_branch),
+            MastNodeType::Loop { body_id: body } => Self::encode_u32_payload(*body),
+            MastNodeType::Block { len } => Self::encode_u32_payload(*len),
+            MastNodeType::Call { callee_id } => Self::encode_u32_payload(*callee_id),
+            MastNodeType::SysCall { callee_id } => Self::encode_u32_payload(*callee_id),
+            MastNodeType::Dyn => [0; 8],
+            MastNodeType::External => [0; 8],
+        }
+    }
+
+    fn encode_join_or_split(left_child_id: u32, right_child_id: u32) -> [u8; 8] {
```

Comment:
The encoding of `JOIN` and `SPLIT` is a bit heavy; I can either document it in a comment here, or in the document that will describe the serialization format of `MastForest` (or both).